kube-runtime/src/reflector/object_ref.rs

```diff
@@ -69,32 +56,41 @@ pub struct Extra {
     pub uid: Option<String>,
 }
 
-impl<K: Resource> ObjectRef<K>
-where
-    K::DynamicType: Default,
-{
+impl ObjectRef {
+    /// Create a blank `ObjectRef` with a name
     #[must_use]
     pub fn new(name: &str) -> Self {
-        Self::new_with(name, Default::default())
+        Self {
+            name: name.into(),
+            namespace: None,
+            types: None,
+            extra: Extra::default(),
+        }
     }
 
+    /// Creates `ObjectRef` from the resource
     #[must_use]
-    pub fn from_obj(obj: &K) -> Self
-    where
-        K: Resource,
-    {
-        Self::from_obj_with(obj, Default::default())
+    pub fn from_obj<K: TypeInfo>(obj: &K) -> Self {
+        let meta = obj.meta();
+        Self {
+            name: meta.name.clone().unwrap_or_default(),
+            namespace: meta.namespace.clone(),
+            types: obj.types(),
+            extra: Extra::from_objectmeta(meta),
+        }
     }
-}
 
-impl<K: Resource> ObjectRef<K> {
+    /// Creates a partial `ObjectRef` from an `OwnerReference`
     #[must_use]
-    pub fn new_with(name: &str, dyntype: K::DynamicType) -> Self {
+    pub fn from_owner(owner: &OwnerReference) -> Self {
         Self {
-            dyntype,
-            name: name.into(),
+            name: owner.name.clone(),
             namespace: None,
-            extra: Extra::default(),
+            types: None,
```

Comment:
`OwnerReference` carries type metadata.

Reply:
Trying to clean up the constructors of `ObjectRef` now. Have grabbed the available information now in `ObjectRef::from_owner` via `OwnerReference` to provide less avenues for not putting valid information in there.

This ctor is used only by `trigger_owners` to create an `ObjectRef`, so using this data here means that it's possible for a user to orphan a controller owned resource by `kubectl edit`ing out the ownerref of a child. Still, that is probably the right behaviour.